evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs

```diff
@@ -568,6 +612,108 @@ impl<'a, 'b, B: Backend, P: PrecompileSet> CheatcodeStackExecutor<'a, 'b, B, P>
         }
     }
 
+    /// Debug execute a transaction by grabbing the debugger from state_mut, adding the runtime.
+    pub fn debug_execute(
+        &mut self,
+        runtime: &mut Runtime,
+        address: Address,
+        code: Rc<Vec<u8>>,
+    ) -> ExitReason {
+        let mut forge_runtime = ForgeRuntime::new_with_runtime(runtime, code);
+        let depth = if let Some(depth) = self.state().metadata().depth() { depth + 1 } else { 0 };
+
+        match self.debug_run(&mut forge_runtime, address, depth) {
+            Capture::Exit(s) => s,
+            Capture::Trap(_) => unreachable!("Trap is Infallible"),
+        }
+    }
+
+    pub fn debug_step(&mut self, runtime: &ForgeRuntime, steps: &mut Vec<DebugStep>) -> bool {
+        let pc = if let Ok(pos) = runtime.inner.machine().position() { *pos } else { 0 };
+        let mut push_bytes = None;
+        if let Some((op, stack)) = runtime.inner.machine().inspect() {
+            let wrapped_op = OpCode(op, None);
+            if let Some(push_size) = wrapped_op.push_size() {
+                let push_start = pc + 1;
+                let push_end = pc + 1 + push_size as usize;
+                if push_end < runtime.code.len() {
+                    push_bytes = Some(runtime.code[push_start..push_end].to_vec());
+                } else {
+                    panic!("PUSH{} exceeds codesize?", push_size)
+                }
+            }
+            let mut stack = stack.data().clone();
+            stack.reverse();
+            steps.push(DebugStep {
+                pc,
+                stack,
+                memory: runtime.inner.machine().memory().clone(),
+                op: wrapped_op,
+                push_bytes,
+            });
+            match op {
+                Opcode::CREATE |
+                Opcode::CREATE2 |
+                Opcode::CALL |
+                Opcode::CALLCODE |
+                Opcode::DELEGATECALL |
+                Opcode::STATICCALL => {
+                    // this would create an interrupt, construct a new vec
+                    true
+                }
+                _ => false,
+            }
+        } else {
+            let mut stack = runtime.inner.machine().stack().data().clone();
+            stack.reverse();
+            steps.push(DebugStep {
+                pc,
+                stack,
+                memory: runtime.inner.machine().memory().clone(),
+                op: OpCode(Opcode::INVALID, None),
+                push_bytes,
+            });
+            true
+        }
+    }
+
+    /// Loop stepping the runtime until it stops.
+    fn debug_run(
+        &mut self,
+        runtime: &mut ForgeRuntime,
+        address: Address,
+        depth: usize,
+    ) -> Capture<ExitReason, ()> {
+        let mut done = false;
+        let mut res = Capture::Exit(ExitReason::Succeed(ExitSucceed::Returned));
+        let mut steps = Vec::new();
+        while !done {
+            if self.debug_step(runtime, &mut steps) {
```

Comment:
We have to add a vector of `steps` *prior* to the call (performed in `runtime.step`) to maintain the call hierarchy and keep the order correct. If we just returned a vector of steps, we would have inner calls *prior* to the parent calls because they complete their execution first. So we have to push the executed steps *first* before the call

pub was a hangover from a refactor, removed